docs/design/sql/08-expressions.md

```diff
@@ -0,0 +1,192 @@
+implementations of `SqlValidator`. Specifically, `SqlValidatorImpl`
+extended by `HazelcastSqlValidator` is used to customize the validation
+logic to align it with our requirements.
+
+As a first step of the validation process, `SqlValidatorImpl` may apply
+some transformations to the passed AST which should not affect the
+semantics of the AST. Such transformations are called "unconditional
+rewrites" (see `SqlValidatorImpl.performUnconditionalRewrites`). Every
+`SqlCall` node has an associated `SqlOperator` returned by its
+`getOperator` method. `SqlOperator` implementations may customize
+various aspects of a certain operator including the rewriting behavior
+itself (see `SqlOperator.rewriteCall`).
+
+Normally, Calcite's parser inspects the hardcoded `SqlStdOperatorTable`
+implementation of `SqlOperatorTable` to resolve a certain operator, so
+it's hard to customize the pre-existing operators. To make the
+customization possible, operators from `SqlStdOperatorTable` are patched
+to the ones from `HazelcastSqlOperatorTable` during the AST rewriting
+process (see `HazelcastSqlValidator.performUnconditionalRewrites` and
+`HazelcastOperatorTableVisitor`).
+
+Usually, semantic validation requires assigning types to every
+expression and its operand expressions, if any. In Calcite, that type
+assignment behavior may be customized in several ways:
+
+- `SqlValidator.deriveType` may customize the behavior globally.
+
+- Each `SqlOperator` implementation may customize its operand type
+  inference strategy (see `SqlOperandTypeInference` and
+  `SqlOperator.getOperandTypeInference`) and its return type inference
+  strategy (see `SqlReturnTypeInference`,
+  `SqlOperator.getReturnTypeInference` and
+  `SqlOperator.inferReturnType`). The first one is used only if some of
+  the operand types are unknown. Also, operand type validation strategy
+  can be customized using an associated `SqlOperandTypeChecker` (see
+  `SqlOperator.getOperandTypeChecker`).
+
+- Additionally, some expressions require their operands to be coerced to
+  a certain type. For instance, `=` operator requires both of its sides
+  to be of the same type. In Calcite, such coercion services are
+  provided by `TypeCoercion` implementations. Specifically,
+  `TypeCoercionImpl` extended by `HazelcastTypeCoercion` is used to
+  customize the coercion logic.
+
+The following general rules apply when assigning types:
+
+- Literals: `TRUE` and `FALSE` literals receive `BOOLEAN` type; numeric
+  literals containing no decimal point (`1`, `42`, etc.) receive the
+  smallest integer type possible (`TINYINT`, `SMALLINT`, `INTEGER` or
+  `BIGINT`); numeric literals containing decimal point (`1.1`, `4.2`,
+  etc.) receive `DECIMAL` type; scientific notation numeric literals
+  (`1e1`, `4.2e2`, etc.) receive `DOUBLE` type; string literals
+  (`'foo'`) receive `VARCHAR` type.
+
+- Parameters: parameter types are inferred from the context: `1.0 + ?`,
```

Comment:
Since the proper parameter inference is still an open question, can we have more examples of how it is actually performed in our product at the moment? For example, comparison and binary arithmetic functions have one inference strategy, while simple functions do it differently.

Reply:
clarified in d1d2929413835f7a462bc88686630dd57203b27c